### PR TITLE
Retrieving known peripherals

### PR DESCRIPTION
**Pull request by indragiek**

### CODE REVIEW

File: ReactiveBLE/RBTCentralManager.m [hunk 1]
```diff
@@ -61,17 +61,14 @@ - (RACSignal *)stateSignal
...
 - (RACSignal *)scanForPeripheralsWithServices:(NSArray *)services options:(NSDictionary *)options
 {
-	return [[[RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
-		RACSerialDisposable *disposable = [[RACSerialDisposable alloc] init];
-		[self.CBScheduler schedule:^{
-			[self.manager scanForPeripheralsWithServices:services options:options];
-			disposable.disposable = [[[self
-				rac_signalForSelector:@selector(centralManager:didDiscoverPeripheral:advertisementData:RSSI:) fromProtocol:@protocol(CBCentralManagerDelegate)]
-				reduceEach:^(CBCentralManager *manager, CBPeripheral *peripheral, NSDictionary *data, NSNumber *RSSI) {
+	return [[[[RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
```
Comment (kastiglione) [inline ReactiveBLE/RBTCentralManager.m]: Block return type can be inferred.

File: ReactiveBLE/RBTCentralManager.m [hunk 2]
```diff
@@ -153,6 +152,46 @@ - (RACSignal *)disconnectPeripheral:(CBPeripheral *)peripheral
...
+		rac_signalForSelector:selector fromProtocol:@protocol(CBCentralManagerDelegate)]
+		take:1]
+		reduceEach:^(CBCentralManager *manager, NSArray *peripherals) {
+			return peripherals;
+		}];
+}
+
+- (RACSignal *)retrievePeripheralsWithIdentifiers:(NSArray *)identifiers
+{
+	return [[[RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
```
Comment (kastiglione) [inline ReactiveBLE/RBTCentralManager.m]: Inferable block return type.

File: ReactiveBLE/RBTCentralManager.m [hunk 3]
```diff
@@ -153,6 +152,46 @@ - (RACSignal *)disconnectPeripheral:(CBPeripheral *)peripheral
...
+		[self.manager retrievePeripheralsWithIdentifiers:identifiers];
+		return disposable;
+	}]
+	subscribeOn:self.CBScheduler]
+	setNameWithFormat:@"RBTCentralManager -retrievePeripheralsWithIdentifiers: %@", identifiers];
+}
+
+- (RACSignal *)retrieveConnectedPeripheralsWithServices:(NSArray *)services
+{
+	return [[[RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
```
Comment (kastiglione) [inline ReactiveBLE/RBTCentralManager.m]: Inferable block return type.

File: ReactiveBLE/RBTCentralManager.m (hunk 2 above)
Comment (kastiglione) [inline ReactiveBLE/RBTCentralManager.m]: For both `-retrievePeripheralsWithIdentifiers:` and `-retrieveConnectedPeripheralsWithServices:`, you could simplify with `+defer:`:

``` objc
[[[RACSignal
    defer:^{
        RACSignal *signal = [self peripheralsSignalForSelector:@selector(centralManager:didRetrievePeripherals:)];
        [self.manager retrievePeripheralsWithIdentifiers:identifiers];
        return signal;
    }]
    subscribeOn:self.CBScheduler]
    setNameWithFormat:…];
```

File: ReactiveBLE/RBTCentralManager.m [hunk 4]
```diff
@@ -153,6 +152,46 @@ - (RACSignal *)disconnectPeripheral:(CBPeripheral *)peripheral
...
+		RACDisposable *disposable = [[self
+			peripheralsSignalForSelector:@selector(centralManager:didRetrieveConnectedPeripherals:)]
+			subscribe:subscriber];
+		
+		[self.manager retrieveConnectedPeripheralsWithServices:services];
+		return disposable;
+	}]
+	subscribeOn:self.CBScheduler]
+	setNameWithFormat:@"RBTCentralManager -retrieveConnectedPeripheralsWithServices: %@", services];
```
Comment (kastiglione) [inline ReactiveBLE/RBTCentralManager.m]: As signals from an object/instance, the format string should conventionally begin with `<%@:%p>`, where the `%@` is fed the object's class.

File: ReactiveBLE/RBTCentralManager.m [hunk 5]
```diff
@@ -153,6 +152,46 @@ - (RACSignal *)disconnectPeripheral:(CBPeripheral *)peripheral
...
+	return [[[RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
+		RACDisposable *disposable = [[self
+			peripheralsSignalForSelector:@selector(centralManager:didRetrievePeripherals:)]
+			subscribe:subscriber];
+		
+		[self.manager retrievePeripheralsWithIdentifiers:identifiers];
+		return disposable;
+	}]
+	subscribeOn:self.CBScheduler]
+	setNameWithFormat:@"RBTCentralManager -retrievePeripheralsWithIdentifiers: %@", identifiers];
```
Comment (kastiglione) [inline ReactiveBLE/RBTCentralManager.m]: `s/RBTCentralManager/<%@:%p>/`

**Comment by indragiek** [comment]
:boom:

File: ReactiveBLE/RBTCentralManager.m [hunk 6]
```diff
@@ -61,17 +61,14 @@ - (RACSignal *)stateSignal
...
 {
-	return [[[RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
-		RACSerialDisposable *disposable = [[RACSerialDisposable alloc] init];
-		[self.CBScheduler schedule:^{
-			[self.manager scanForPeripheralsWithServices:services options:options];
-			disposable.disposable = [[[self
-				rac_signalForSelector:@selector(centralManager:didDiscoverPeripheral:advertisementData:RSSI:) fromProtocol:@protocol(CBCentralManagerDelegate)]
-				reduceEach:^(CBCentralManager *manager, CBPeripheral *peripheral, NSDictionary *data, NSNumber *RSSI) {
+	return [[[[RACSignal createSignal:^(id<RACSubscriber> subscriber) {
+		[self.manager scanForPeripheralsWithServices:services options:options];
```
Comment (kastiglione) [inline ReactiveBLE/RBTCentralManager.m]: In the methods below, you get the delegate signal with `-rac_signalForSelector:` before calling a method on `manager`. In this case, `-scanForPeripheralsWithServices:options:` is called before `-rac_signalForSelector:`, which could theoretically result in a missed value on the signal. Should the order be swapped?

Comment (indragiek) [inline ReactiveBLE/RBTCentralManager.m]: Good catch. I swear I fixed that a while ago but it must've gotten messed up again during a merge. Fixed.

**Comment by kastiglione** [comment]
:gem:

**Pull request merged**